### PR TITLE
Attempt to render accessibility render information using Android Views

### PR DESCRIPTION
**Pull request by geoff-powell**
Since render extensions use of fonts with Java's Graphics2D apis scale differently per OS, this attempts to render the accessibility information using Android Views instead.

Would love some feedback 👍🏽 

[This change](https://github.com/cashapp/paparazzi/pull/318/files?short_path=6b66eaa#diff-6b66eaa31c0052078faae23cfce37fde9a6fb4b8f59a269a1f6ff68161c05341) shows the new output with Android Views.

### CODE REVIEW

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt [hunk 1]
```diff
@@ -77,7 +85,7 @@ class AccessibilityRenderExtensionTest {
             relativePath = expected.path,
             image = image,
             goldenImage = ImageIO.read(expected),
-            maxPercentDifferent = 0.2,
+            maxPercentDifferent = 0.1, // Default percent defined in paparazzi.
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt:88]: can this be set to 0 now?

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt [hunk 2]
```diff
@@ -20,14 +21,21 @@ import javax.imageio.ImageIO
...
   @Test
   fun test() {
     val view = buildView(paparazzi.context)
-    paparazzi.snapshot(view)
+
+    val deviceConfig = DeviceConfig.NEXUS_5.copy(
+      // Needed to render accessibility content next to main content.
+      screenWidth = DeviceConfig.NEXUS_5.screenWidth * 2,
+      softButtons = false
+    )
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt:37]: inline to line 24

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt [hunk 3]
```diff
@@ -20,14 +21,21 @@ import javax.imageio.ImageIO
...
+    paparazzi.snapshot(view, deviceConfig = deviceConfig)
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt:38]: rely on rule's deviceConfig per comment above?

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt [hunk 4]
```diff
@@ -20,14 +21,21 @@ import javax.imageio.ImageIO
 class AccessibilityRenderExtensionTest {
   @get:Rule
   val paparazzi = Paparazzi(
+    deviceConfig = DeviceConfig.NEXUS_5,
     snapshotHandler = TestSnapshotVerifier(),
-    renderExtensions = setOf(AccessibilityRenderExtension())
+    renderExtensions = setOf(AccessibilityRenderExtension()),
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt:26]:
```suggestion
    renderExtensions = setOf(AccessibilityRenderExtension())
```

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 5]
```diff
@@ -69,7 +69,7 @@ class Paparazzi(
   private val appCompatEnabled: Boolean = true,
   private val maxPercentDifference: Double = 0.1,
   private val snapshotHandler: SnapshotHandler = determineHandler(maxPercentDifference),
-  private val renderExtensions: Set<RenderExtension> = setOf()
+  private val renderExtensions: Set<RenderExtension> = setOf(),
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:72]:
```suggestion
  private val renderExtensions: Set<RenderExtension> = setOf()
```

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 6]
```diff
@@ -238,12 +238,20 @@ class Paparazzi(
...
         withTime(0L) {
           // Initialize the choreographer at time=0.
         }
 
-        viewGroup.addView(view)
+        if (renderExtensions.isNotEmpty()) {
+          viewToRender = renderExtensions.fold(view) { view, renderExtension ->
+            return@fold renderExtension.renderView(view)
+          }
+        }
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:252]:
```suggestion
       val modifiedView = renderExtensions.fold(view) { view, renderExtension ->
          renderExtension.renderView(view)
       }
```

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 7]
```diff
@@ -238,12 +238,20 @@ class Paparazzi(
     val frameHandler = snapshotHandler.newFrameHandler(snapshot, frameCount, fps)
     frameHandler.use {
       val viewGroup = bridgeRenderSession.rootViews[0].viewObject as ViewGroup
+      var viewToRender = view
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:241]:
```suggestion
```

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 8]
```diff
@@ -254,13 +262,13 @@ class Paparazzi(
 
             var image = bridgeRenderSession.image
             renderExtensions.forEach {
-              image = it.render(snapshot, view, image)
+              image = it.postRender(view, image)
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:265]: discussed offline.  would be nice to also render the view bound color rects using view hierarchy manipulation vs post-image rendering.

**Review by jrodbx — CHANGES_REQUESTED** [review]
1st round of feedback.  overall looks great!

**Comment by geoff-powell** [comment]
> discussed offline. would be nice to also render the view bound color rects using view hierarchy manipulation vs post-image rendering.

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt (hunk 1 above)
Comment (geoff-powell) [reply paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt]: unfortunately there is a small diff between what mac generates and what is generated on windows. ([build](https://github.com/cashapp/paparazzi/actions/runs/1477112544))

delta-accessibility

For now I will leave this at 0.1

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt [hunk 9]
```diff
@@ -15,97 +15,158 @@
...
-
-  // We have seen issues with the built in font rendering incorrectly on different OS (Linux, Mac OS, etc.).
-  // We will use a roboto font provided by layoutLib to render accessibility metadata.
-  private val robotoFont: Font by lazy {
-    val inputStream = SystemFonts.getAvailableFonts().first { it.file.name.contains("Roboto-Medium.ttf") }.file.inputStream()
-    inputStream.use {
-      Font.createFont(Font.TRUETYPE_FONT, it)
+
+  override fun renderView(
+    contentView: View,
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt:38]:
```suggestion
    contentView: View
```

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt [hunk 10]
```diff
@@ -15,97 +15,158 @@
...
+    TypedValue.COMPLEX_UNIT_DIP,
+    value,
+    resources.displayMetrics
+  )
 
-  private fun BufferedImage.margin(): Int = height / 80
-}
\ No newline at end of file
+private fun View.dip(value: Int): Int =
+  dip(value.toFloat()).toInt()
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt:172]:
```suggestion
private fun View.dip(value: Int): Int = dip(value.toFloat()).toInt()
```

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt [hunk 11]
```diff
@@ -15,97 +15,158 @@
...
 
-      graphics.drawString(
-        accessibilityText,
-        drawStartX + colorRectSize + image.margin(),
-        drawStartY + colorRectSize / 3 + graphics.fontMetrics.height / 2
-      )
-    }
-    return canvasImage
-  }
+      fun buildLayoutParams(
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt:112]: inline

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt [hunk 12]
```diff
@@ -15,97 +15,158 @@
...
+    }
 
-      graphics.color = getColor(view)
-      graphics.fillRoundRect(
-        drawStartX, drawStartY, colorRectSize, colorRectSize,
-        colorRectSize / 4, colorRectSize / 4
-      )
-      graphics.color = DEFAULT_TEXT_COLOR
-      val accessibilityText = view.iterableTextForAccessibility.toString()
+    fun buildAccessibilityRow(view: View, iterableTextForAccessibility: CharSequence): View {
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt:106]: intentionally local method?

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/RenderSettings.kt [hunk 13]
```diff
@@ -55,7 +58,7 @@ internal object RenderSettings {
     return DEFAULT_RENDER_COLORS[colorIndex]
   }
 
-  private fun Color.withAlpha(alpha: Int): Color {
+  fun Color.withAlpha(alpha: Int): Color {
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/RenderSettings.kt:61]:
```suggestion
  internal fun Color.withAlpha(alpha: Int): Color {
```

File: paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt [hunk 14]
```diff
@@ -53,6 +61,16 @@ class AccessibilityRenderExtensionTest {
         contentDescription = "Margin Sample"
       })
 
+      addView(View(context).apply {
+        layoutParams = LinearLayout.LayoutParams(100, 100).apply {
+          setMarginsRelative(20, 20, 20, 20)
+        }
+        foreground = GradientDrawable(GradientDrawable.Orientation.TL_BR, intArrayOf(Color.YELLOW, Color.BLUE)).apply {
```
Comment (jrodbx) [inline paparazzi/src/test/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtensionTest.kt:68]: imports?

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/RenderSettings.kt [hunk 15]
```diff
@@ -46,6 +46,9 @@ internal object RenderSettings {
     }
   }
 
+  fun Color.toColorInt(): Int =
+    android.graphics.Color.argb(alpha, red, green, blue)
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/accessibility/RenderSettings.kt:50]:
```suggestion
  internal fun Color.toColorInt(): Int = android.graphics.Color.argb(alpha, red, green, blue)
```

Comment (jrodbx) [reply paparazzi/src/main/java/app/cash/paparazzi/accessibility/RenderSettings.kt:50]: also, group next to Color.withAlpha?

File: paparazzi/src/main/java/app/cash/paparazzi/RenderExtension.kt [hunk 16]
```diff
@@ -22,5 +22,7 @@ import java.awt.image.BufferedImage
  * An extension for overlaying additional information on top of each rendered frame.
  */
 interface RenderExtension {
-  fun render(snapshot: Snapshot, view: View, image: BufferedImage): BufferedImage
+  // Allows the extension to add its own view hierarchy to the `contentView`.
+  // This is called before anything is rendered by Paparazzi
+  fun renderView(contentView: View): View
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/RenderExtension.kt:27]:
```suggestion
  /**
   * Allows this extension to modify the view hierarchy represented by [contentView].
   *
   * Returns the root view of the modified hierarchy.
   */
  fun render(contentView: View): View
```

**Review by jrodbx — APPROVED** [review]
Some minor nits.  Nice improvement!

File: paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt (hunk 12 above)
Comment (geoff-powell) [reply paparazzi/src/main/java/app/cash/paparazzi/accessibility/AccessibilityRenderExtension.kt:106]: nope, I will move it

**Pull request merged**